### PR TITLE
Changes to get a particular generator working in an example project

### PR DESCRIPTION
**Pull request by token-cjg**
@HungryProton @fire 

Sorry that this is a bit messy, but I'm not very experienced in contributing to open source projects.

(Note: Pull request against the develop branch, not master.)

(Further Note: This was developed against Godot 3.4rc1.)

### Description

Enabling changes to allow fence.tpgn [in this project](https://github.com/token-cjg/TestProtongraphProject) to generate a fence along a curve when invoking this application, also when compiled locally.

A few things have been modified in this change:

* Alters the way resources are handled, so that resources are passed to the generator from the client by reference (as resource_references) before being passed back to the client for rendering.
* Some changes to the public_api and in a number of other places to get things working.
* Comments out some C++ code, this was required in order to be able to build for OSX.  See the issue here https://github.com/protongraph/protongraph/issues/100 for more context.

Note that these changes were worked on in parallel with changes to the legacy ConceptGraph code and the Sync-Godot addon manually added to this example project: https://github.com/token-cjg/TestProtongraphProject.

### Testing Instructions
* Run ./godot.3.4rc1 to run Protongraph with these changes mounted
* Open the example project https://github.com/token-cjg/TestProtongraphProject in Godot 3.4rc1.
* Ensure that you activate the ProtonGraph Sync plugin within "assetlib->addons" in Godot 3.4rc1.
* Open scenes/TestProtongraph.tscn.
* Click on the "Fence" node in the scenetree
* Make sure that the example project is connected to Protongraph via sync-godot (you should see a green check in the Node details UI)
* Click "Rebuild" in the Node details UI for the Fence node in the TestProtongraph scene within the example project.
* Wait a fraction of a second for the data in the Input subnode of the Fence object to be compiled by Sync-Godot, serialized, sent to Protongraph, deserialized, worked on, serialized and sent back via the WebSocket connection to the waiting Sync-Godot module in the Test project, deserialized, interpreted, and rendered.
* You should see a fence generated as a sequence of output nodes.

### References
* Youtube demonstration [here](https://www.youtube.com/watch?v=xN11INQlro8&ab_channel=ChrisGoddard)
* Example project that is required in order for this to work: https://github.com/token-cjg/TestProtongraphProject

### Demo (click on the image to view)

[Kicking the tires of Protongraph](https://www.youtube.com/watch?v=xN11INQlro8&ab_channel=ChrisGoddard)

### CODE REVIEW

**Comment by HungryProton** [comment]
Thanks for the PR! I'll have to take a look at this after work.

**Comment by HungryProton** [comment]
Looks good to me!
The only thing preventing me to merge this is the binaries. It would be great if you could find a way to completely remove them from the history (not just a git rm, as they would still appear in older commits).
There's the two Godot binaries and the content of the build folder that take almost 300mb of space.

**Comment by token-cjg** [comment]
That is great to hear! 🥳 

Sure thing, after work today I'll make local copies of the files I still need, delete the files of which you speak, and then squash all the commits together in my repo to decrease the size of the requested change.

File: common/autoloads/remote_manager.gd [hunk 1]
```diff
@@ -28,22 +28,58 @@ func _set_inspector_values(tpl: Template, values: Array) -> void:
...
+			for resource_reference in resource_references:
+				#print("current resource_reference ...")
+				#print(resource_reference)
+				if resource_reference && resource_reference["name"] == input.name && resource_reference["resource_path"]:
+					#print(resource)
+					#print(input)
+					#print(child_transversal)
+					tpl.set_remote_resource(input.name, child_transversal + [resource_reference.name], resource_reference["resource_path"])
+				else:
+					# TODO: generalise to multiple resource_references as children of a particular top-level input.
```
Comment (fire) [inline common/autoloads/remote_manager.gd:62]: Are you planning to resolve this todo? I would remove the comment if it is not planned.

File: core/template.gd [hunk 2]
```diff
@@ -365,10 +366,27 @@ func get_remote_input(name: String):
 		return _remote_inputs[name]
 	return null
 
+func get_remote_resource(name: String):
+	#print("in the get_remote_resource function")
```
Comment (fire) [inline core/template.gd:370]: Print statements

File: core/template.gd [hunk 3]
```diff
@@ -394,15 +412,34 @@ func _run_generation() -> void:
...
+			call_deferred("emit_signal", "thread_completed")
+			return
+	
+	for remote_input in _remote_inputs:
+		#print("setting data in input_node")
+		#print(remote_input)
+		if not remote_input:
+			_remote_inputs.erase(remote_input)
+			continue
+		# todo handle setting data in _input_nodes
```
Comment (fire) [inline core/template.gd:427]: More todos statements. Remove or fix?

**Comment by HungryProton** [comment]
Thanks for the PR
Once I'm done with the 4.0 port / refactor, I'll cherry-pick the commit into the new codebase.

**Pull request merged**